### PR TITLE
Android 10+ View.setSystemGestureExclusionRects

### PR DESCRIPTION
**Pull request by shihaohong**
## Description
Adds a platform channel that allows the Flutter framework to [set system gesture exclusion rects](https://developer.android.com/reference/android/view/View.html#setSystemGestureExclusionRects(java.util.List%3Candroid.graphics.Rect%3E)) for Android Q+. 

The corresponding framework branch used to test this behavior is https://github.com/shihaohong/flutter/tree/set-exclusion-rects.

## Demo
Demo (exclude top-left corner of app):
resized-gif

Sample code: 
```
                SystemGestureExclusionRects.setSystemGestureExclusionRects([
                  {
                    "top": 0,
                    "right": 500,
                    "bottom": 500,
                    "left": 0,
                  },
                ]);
```

## Related Issues
Related to https://github.com/flutter/flutter/issues/32748

## Testing
Testing is blocked for Android SDK Q+ (29+). Will track for follow up updates to unit tests on https://github.com/flutter/flutter/issues/38626

### CODE REVIEW

**Comment by PieterAelse** [comment]
There's a limit of 200dp for exclusion height: https://android-developers.googleblog.com/2019/08/final-beta-update-official-android-q.html 

> We've made further refinements to Gesture Navigation in Beta 6 based on user feedback. First, to ensure reliable and consistent operation, there's a 200dp vertical app exclusion limit for the Back gesture.

Will there be anything on the Flutter side for this? I don't think Google is throwing an exception or anything, but the least thing possible is mention it in the Flutter method documentation.
Also, as far as I can see the 500 from the example is pixels and not dp?

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java [hunk 1]
```diff
@@ -581,6 +503,149 @@ public SystemChromeStyle(
...
 
+  /**
+   * A handler of incoming platform channel method calls received from Flutter.
+   * It first determines the platform's API to be called. If it exists, it then
+   * decodes incoming arguments, if needed, into a format that is necessary for the API.
+   */
+  @VisibleForTesting
+  protected class PlatformMethodCallHandler implements MethodChannel.MethodCallHandler {
+    @Override
+    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
```
Comment (mklim) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java]: I wish the diff were better here. I spot checked it, assuming that nothing in this was changed from the original besides the new method call.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java [hunk 2]
```diff
@@ -581,6 +503,149 @@ public SystemChromeStyle(
...
+            result.success(null);
+            break;
+          case "SystemGestures.setSystemGestureExclusionRects":
+            if (!(arguments instanceof JSONArray)) {
+              String inputTypeError = "Input type is incorrect. Ensure that a List<Map<String, int>> is passed as the input for SystemGestureExclusionRects.setSystemGestureExclusionRects.";
+              result.error("inputTypeError", inputTypeError, null);
+              break;
+            }
+
+            Log.v("Got here", arguments.toString());
```
Comment (mklim) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java]: This log probably isn't useful and should be removed, even at the `verbose` level.

File: shell/platform/android/test/README.md [hunk 3]
```diff
@@ -18,7 +18,7 @@ integration tests in other repos.
 3. Add your class to the `@SuiteClasses` annotation in `FlutterTestSuite.java`.
    This makes sure the test is actually executed at run time.
 4. Write your test.
-5. Build and run with `testing/run_tests.py [--type=java] [filter=<test_class_name>]`.
+5. Build and run with `testing/run_tests.py [--type=java] [--java-filter=<test_class_name>]`.
```
Comment (mklim) [inline shell/platform/android/test/README.md]: Thanks for updating this!

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java (hunk 2 above)
Comment (shihaohong) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java]: This was an artifact from when I was testing my tests. I'll remove it!

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java [hunk 4]
```diff
@@ -32,129 +34,9 @@
   public final MethodChannel channel;
   @Nullable
   private PlatformMessageHandler platformMessageHandler;
-
-  private final MethodChannel.MethodCallHandler parsingMethodCallHandler = new MethodChannel.MethodCallHandler() {
```
Comment (matthew-carroll) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java]: @shihaohong why was this moved?

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java [hunk 5]
```diff
@@ -581,6 +503,147 @@ public SystemChromeStyle(
     }
   }
 
+  /**
+   * A handler of incoming platform channel method calls received from Flutter.
+   * It first determines the platform's API to be called. If it exists, it then
+   * decodes incoming arguments, if needed, into a format that is necessary for the API.
+   */
+  @VisibleForTesting
+  protected class PlatformMethodCallHandler implements MethodChannel.MethodCallHandler {
```
Comment (matthew-carroll) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java]: I'd like for us to avoid non-static inner classes. I think in this case I'd like for us to revert the movement of this handler definition back to where it was. But in future cases where we do define inner classes, they should be static to avoid the possibility of accidentally leaking references to the outer class.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/PlatformChannelTest.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@Config(manifest=Config.NONE)
+@RunWith(RobolectricTestRunner.class)
+public class PlatformChannelTest {
+    @Test
+    public void setSystemExclusionRectsSendsSuccessMessageToFramework() throws JSONException {
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/systemchannels/PlatformChannelTest.java]: Please name all tests as `itDoesSomethingWhenSomethingHappens` or as close to that as possible. This helps ensure that we writing unit tests here instead of integration tests.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/PlatformChannelTest.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        platformChannel.parsingMethodCallHandler.onMethodCall(callSystemGestureExclusionRects, resultsMock);
+        verify(resultsMock, times(1)).error(
+            "error",
+            "JSON error: Incorrect JSON data shape. To set system gesture exclusion rects, \n" +
+            "a JSONObject with top, right, bottom and left values need to be set to int values.",
+            null
+        );
+    }
+
+    private class ResultsMock implements Result {
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/systemchannels/PlatformChannelTest.java]: Please make all inner classes static to avoid hidden references that can cause leaks and complications.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/PlatformChannelTest.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@Config(manifest=Config.NONE)
+@RunWith(RobolectricTestRunner.class)
+public class PlatformChannelTest {
+    @Test
+    public void setSystemExclusionRectsSendsSuccessMessageToFramework() throws JSONException {
+        DartExecutor dartExecutor = mock(DartExecutor.class);
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/systemchannels/PlatformChannelTest.java]: Please denote with comments in each test the part that is setup vs execution vs verification so that developers can more easily check the part they're interested in.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/PlatformChannelTest.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        PlatformChannel platformChannel = new PlatformChannel(dartExecutor);
+        PlatformMessageHandler platformMessageHandler = mock(PlatformMessageHandler.class);
+        platformChannel.setPlatformMessageHandler(platformMessageHandler);
+
+        int top = 0;
+        int right = 500;
+        int bottom = 250;
+        int left = 0;
+
+        ResultsMock resultsMock = mock(ResultsMock.class);
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/systemchannels/PlatformChannelTest.java]: Can't you just `mock(Result.class)` instead of introducing `ResultsMock`? Or does Mockito not allow that?

**Comment by shihaohong** [comment]
@matthew-carroll How should I address your review suggestions? Should I make a separate PR addressing them?

**Comment by matthew-carroll** [comment]
Yeah, I think that would be fine.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java (hunk 5 above)
Comment (shihaohong) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java]: I had a discussion with @mklim and we felt that it was difficult to read an inline class definition within another class. I'll revert this in a subsequent PR
